test/functional/system/blocks/chain/apply_block.js

```diff
@@ -0,0 +1,453 @@
+									eachCb();
+								});
+						},
+						done
+					);
+				});
+			});
+
+			describe('after applying new block passes', () => {
+				beforeEach(done => {
+					library.modules.blocks.chain.applyBlock(block, true, done);
+				});
+
+				it('should applyUnconfirmedStep for block transactions', done => {
+					async.forEach(
+						[blockAccount1, blockAccount2],
+						(account, eachCb) => {
+							localCommon
+								.getAccountFromDb(library, account.address)
+								.then(accountRow => {
+									expect(accountRow.mem_accounts.u_username).to.equal(
+										account.username
+									);
+									expect(accountRow.mem_accounts.u_isDelegate).to.equal(1);
+									eachCb();
+								});
+						},
+						done
+					);
+				});
+			});
+		});
+
+		describe('applyConfirmedStep', () => {
+			const randomUsername = randomUtil.username();
+			describe('after applying new block fails', () => {
+				beforeEach(done => {
+					// Making mem_account invalid
+					library.logic.account.set(
+						blockAccount1.address,
+						{
+							isDelegate: 1,
+							username: randomUsername,
+							publicKey: blockTransaction1.senderPublicKey,
+						},
+						err => {
+							expect(err).to.not.exist;
+							library.modules.blocks.chain.applyBlock(block, true, () =>
+								done()
+							);
+						}
+					);
+				});
+
+				it('should have pooled transactions in queued state', done => {
+					async.forEach(
+						[poolAccount3, poolAccount4],
+						(account, eachCb) => {
+							localCommon
+								.getAccountFromDb(library, account.address)
+								.then(accountRow => {
+									expect(0).to.equal(accountRow.mem_accounts.u_secondSignature);
+									eachCb();
+								});
+						},
+						done
+					);
+				});
+
+				// Should be unskipped once transaction
+				it.skip('should revert applyconfirmedStep on block transactions', done => {
+					async.forEach(
+						[blockAccount1, blockAccount2],
+						(account, eachCb) => {
+							localCommon
+								.getAccountFromDb(library, account.address)
+								.then(accountRow => {
+									// the transaction will fail, so we will have the username, isDelegate we initially set
+									if (account == blockAccount1) {
+										expect(accountRow.mem_accounts.username).to.equal(
+											randomUsername
+										);
+										expect(accountRow.mem_accounts.isDelegate).to.equal(1);
+									}
+
+									if (account == blockAccount2) {
+										expect(accountRow.mem_accounts.username).to.eql(null);
+										expect(accountRow.mem_accounts.isDelegate).to.equal(0);
+									}
+
+									expect(accountRow.mem_accounts.u_username).to.equal(null);
+									expect(accountRow.mem_accounts.u_isDelegate).to.equal(0);
+									eachCb();
+								});
+						},
+						done
+					);
+				});
+			});
+
+			describe('after applying a new block', () => {
+				beforeEach(done => {
+					library.modules.blocks.chain.applyBlock(block, true, done);
```

Comment:
@SargeKhan  We are directly calling the unit of code and testing impact and response for that particular unit of code. Should not it be a unit test? I don't see this `applyBlock` method is relying on any thing outside of its call so wee have to initiate the full application for it.

Reply:
This function is internally calling transactions module, transaction.apply, transaction.undo and calling the database as well. So, that's why I am using functional system tests.